experimental/Rings/WeierstrassPreparationTheorem.jl

```diff
@@ -0,0 +1,240 @@
+
+compute for a given polynomial `f` an integer `d` such that `f` is `x_n`-general of order `d` 
+with `x_n` the n-th variable of the polynomial ring of `f` 
+"""
+function order_of_generality(f::MPolyElem, n::Int)
+	 
+	vect = fill(parent(f)(0), nvars(parent(f))) 
+	vect[n] = gens(parent(f))[n] 
+	f = evaluate(f, vect) 
+	f != 0 || return false, -1
+	return true, degree(leading_monomial(f, ordering=neglex(gens(parent(f)))), n)
+end
+
+"""
+`jet(f::MPolyElem, degree_f::Int)`
+
+Given a polynomial `f` and an integer degree `degree_f` , this function returns the polynomial,
+which contains only the monomials of `f`, which total degree is lower than or equal to `degree_f`
+"""
+function jet(f::MPolyElem, degree_f::Int) 
+	while f != 0
+		if total_degree(leading_monomial(f, ordering=deglex(gens(parent(f))))) > degree_f
+			f = f-coeff(f, leading_monomial(f, ordering=deglex(gens(parent(f)))))*leading_monomial(f, ordering=deglex(gens(parent(f))))
+		else 
+			break
+		end
+	end
+	return f
+end
+
+
+"""
+`jet(f::MPolyElem, degree_f::Int, vecw::Vector{Int})` 
+
+Given a polynomial `f`, an integer degree `degree_f` and an integer vector `vecw`, which has to have 
+the same length as length(parent(f)), this function returns a polynomial, which only contains 
+monomial, whose exponents taken as a vector multiplied (scalar product) with the given 
+weight-vector `intw` are lower than or equal to `degree_f`
+"""
+function jet(f::MPolyElem, degree_f::Int, vecw::Vector{Int}) 
+
+	vect = collect(exponent_vectors(f))
+	i = length(f)
+	while i >= 1
+		if transpose(vecw)*vect[i] > degree_f
+			f = f-term(f, i)
+		end
+		i = i-1	
+	end
+	return f
+end
+
+"""
+`invert_unit(u::MPolyElem, n::Int)`
+If `u` is an unit, returning boolean value for success and the inverse of `u`. 
+Else return false and the zero polynomial. 
+"""
+function invert_unit(u::MPolyElem, n::Int)
+	if u != 0 && total_degree(leading_monomial(u, ordering=negdegrevlex(gens(parent(u))))) == 0
+		u0 = jet(u, 0)
+		u = jet(1-divexact(u, u0), n)
+		ui = u
+		v = 1+u
+		if u != 0
+			i = div(n, total_degree(leading_monomial(u, ordering=negdegrevlex(gens(parent(u))))))
+			while i > 1
+				ui = jet(ui*u, n)
+				v = v+ui
+				i = i-1
+			end
+		end
+		v = divexact(jet(v, n), u0)
+		return true, v
+	else
+		return false, parent(u)(0)
+	end
+end
+
+"""
+`last_var_general(f::MPolyElem, n::Int)`
+
+transform `f` into a polynomial, which is x_n general of finite order for x_n the `n`-th ring variable.
+"""
+var_general(f::MPolyElem, n::Int) = var_general(f, n, 0)
+
+"""
+`var_general(f::MPolyElem, n::Int, nesting_depth::Int)`
+
+transform `f` into a polynomial, which is x_n general of finite order for x_n the `n`-th ring variable.
+`nesting_depth` is the counter of recursive calls. First call with `nesting_depth` = 0
+"""
+function var_general(f::MPolyElem, n::Int, nesting_depth::Int) 
+	R = parent(f)
+	j = nvars(R)
+	_, d = order_of_generality(f, n) 
+	if d > 0 
+		return f	
+	else 
```

Comment:
The more indented a function, the harder it gets to understand it. Here, you can save one level: just do
```suggestion
	end
```
and then de-indent the following code, and finally drop the closing `end`